Running the controller locally fails on Apple silicon

Running `make run` on Apple M1 Max fails with:

```console
$ LIBGIT2_FORCE=1 make run
cd api; /Users/stefanprodan/go/bin/controller-gen object:headerFile="../hack/boilerplate.go.txt" paths="./..."
go fmt ./...
cd api; go fmt ./...
PKG_CONFIG_PATH=/Users/stefanprodan/go/src/github.com/fluxcd/source-controller/hack/libgit2/lib/pkgconfig \
go vet ./...
cd api; go vet ./...
/Users/stefanprodan/go/bin/controller-gen crd:crdVersions=v1 rbac:roleName=manager-role paths="./..." output:crd:artifacts:config="config/crd/bases"
cd api; /Users/stefanprodan/go/bin/controller-gen crd:crdVersions=v1 rbac:roleName=manager-role paths="./..." output:crd:artifacts:config="../config/crd/bases"
LD_LIBRARY_PATH=/Users/stefanprodan/go/src/github.com/fluxcd/source-controller/hack/libgit2/lib \
go run ./main.go
dyld[77871]: Library not loaded: @rpath/libgit2.1.1.dylib
  Referenced from: /private/var/folders/fm/v67j6c7n1k5g46m38yb15yvm0000gn/T/go-build707344674/b001/exe/main
  Reason: tried: '/usr/local/lib/libgit2.1.1.dylib' (no such file), '/usr/lib/libgit2.1.1.dylib' (no such file)
signal: abort trap
make: *** [Makefile:70: run] Error 1
```

Configuration:

```console
$ cat ~/.bash_profile
export PATH="/opt/homebrew/opt/make/libexec/gnubin:$PATH"
export PATH="/opt/homebrew/opt/openssl@1.1/bin:$PATH"
export LDFLAGS="-L/opt/homebrew/opt/openssl@1.1/lib"
export CPPFLAGS="-I/opt/homebrew/opt/openssl@1.1/include"
export PKG_CONFIG_PATH="/opt/homebrew/opt/openssl@1.1/lib/pkgconfig"

$ openssl version
OpenSSL 1.1.1l  24 Aug 2021

$ go version
go version go1.16.10 darwin/arm64

$ uname -av
Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:01 PDT 2021; root:xnu-8019.41.5~1/RELEASE_ARM64_T6000 arm64
```

I've found no way to run the controller locally on Apple M1, my current workaround is to use Docker for Mac:

```bash
# create a linux/arm64 cluster
kind create cluster

# install Flux using linux/arm64 images
flux install

# build source-controller linux/arm64 image
BUILD_PLATFORMS=linux/arm64 BUILD_ARGS=--load make docker-build

# load the image in kind
kind load docker-image fluxcd/source-controller:latest

# set the image in-cluster
kubectl -n flux-system set image deployment/source-controller manager=fluxcd/source-controller:latest

# verify it works
flux check
```

--- a/Makefile
+++ b/Makefile
@@ -39,7 +39,12 @@ endif
 
 ifeq ($(shell uname -s),Darwin)
 	LIBGIT2 := $(LIBGIT2_LIB_PATH)/libgit2.$(LIBGIT2_VERSION).dylib
+	HAS_BREW := $(shell brew --version 2>/dev/null)
+ifdef HAS_BREW
+	HAS_OPENSSL := $(shell brew --prefix openssl@1.1)
 endif
+endif
+
 
 # API (doc) generation utilities
 CONTROLLER_GEN_VERSION ?= v0.5.0
@@ -52,23 +57,53 @@ else
 GOBIN=$(shell go env GOBIN)
 endif
 
+ifeq ($(strip ${PKG_CONFIG_PATH}),)
+	MAKE_PKG_CONFIG_PATH = $(LIBGIT2_LIB_PATH)/pkgconfig
+else
+	MAKE_PKG_CONFIG_PATH = ${PKG_CONFIG_PATH}:$(LIBGIT2_LIB_PATH)/pkgconfig
+endif
+
+ifdef HAS_OPENSSL
+	MAKE_PKG_CONFIG_PATH := $(MAKE_PKG_CONFIG_PATH):$(HAS_OPENSSL)/lib/pkgconfig
+endif
+
 all: build
 
 build: $(LIBGIT2) ## Build manager binary
-	PKG_CONFIG_PATH=$(LIBGIT2_LIB_PATH)/pkgconfig/ \
+ifeq ($(shell uname -s),Darwin)
+	PKG_CONFIG_PATH=$(MAKE_PKG_CONFIG_PATH) \
+	CGO_LDFLAGS="-Wl,-rpath,$(LIBGIT2_LIB_PATH)" \
 	go build -o bin/manager main.go
+else
+	PKG_CONFIG_PATH=$(MAKE_PKG_CONFIG_PATH) \
+	go build -o bin/manager main.go
+endif
 
 test: $(LIBGIT2) test-api  ## Run tests
+ifeq ($(shell uname -s),Darwin)
 	LD_LIBRARY_PATH=$(LIBGIT2_LIB_PATH) \
-	PKG_CONFIG_PATH=$(LIBGIT2_LIB_PATH)/pkgconfig/ \
+	PKG_CONFIG_PATH=$(MAKE_PKG_CONFIG_PATH) \
+	CGO_LDFLAGS="-Wl,-rpath,$(LIBGIT2_LIB_PATH)" \
 	go test ./... -coverprofile cover.out
+else
+	LD_LIBRARY_PATH=$(LIBGIT2_LIB_PATH) \
+	PKG_CONFIG_PATH=$(MAKE_PKG_CONFIG_PATH) \
+	go test ./... -coverprofile cover.out
+endif
 
 test-api: ## Run api tests
 	cd api; go test ./... -coverprofile cover.out
 
 run: $(LIBGIT2) generate fmt vet manifests  ## Run against the configured Kubernetes cluster in ~/.kube/config
+ifeq ($(shell uname -s),Darwin)
+	LD_LIBRARY_PATH=$(LIBGIT2_LIB_PATH) \
+	CGO_LDFLAGS="-Wl,-rpath,$(LIBGIT2_LIB_PATH)" \
+	go run ./main.go
+else
 	LD_LIBRARY_PATH=$(LIBGIT2_LIB_PATH) \
 	go run ./main.go
+endif
+
 
 install: manifests  ## Install CRDs into a cluster
 	kustomize build config/crd | kubectl apply -f -
@@ -102,9 +137,16 @@ fmt:  ## Run go fmt against code
 	cd api; go fmt ./...
 
 vet: $(LIBGIT2)	## Run go vet against code
-	PKG_CONFIG_PATH=$(LIBGIT2_LIB_PATH)/pkgconfig \
+ifeq ($(shell uname -s),Darwin)
+	PKG_CONFIG_PATH=$(MAKE_PKG_CONFIG_PATH) \
+	CGO_LDFLAGS="-Wl,-rpath,$(LIBGIT2_LIB_PATH)" \
 	go vet ./...
 	cd api; go vet ./...
+else
+	PKG_CONFIG_PATH=$(MAKE_PKG_CONFIG_PATH) \
+	go vet ./...
+	cd api; go vet ./...
+endif
 
 generate: controller-gen  ## Generate API code
 	cd api; $(CONTROLLER_GEN) object:headerFile="../hack/boilerplate.go.txt" paths="./..."

